### PR TITLE
Refactor Device Node Implementation in SchemaRegion

### PR DESCRIPTION
**Pull request by MarcosZyk**
### Motivation

The original implementation of device node in SchemaRegion is based on independent classes, DeviceMNode and DatabaseDeviceMNode. 

When a normal internal node should be upgraded to a device node or a device node should be degraded to a normal internal node, node replacement shall happens.

This manner brings huge burden to concurrency control and code maintenance.

For example:

```
Thread A                                    Thread B
create      node  root.db.a
                                                   upgrade  node  root.db.a
create      node  root.db.a.b
                                                   create      leaf    root.db.a.d
upgrade  node  root.db.a.b

create      leaf    root.db.a.b.c
```
After thread B upgrade the node, the instance hold by thread A is no longer the one actually on the tree.

### Modification

The new implementation use a field, deviceNodeInfo, in internal node to represent whether this instance is a device node.

When the node should be processed as a device node, a temporary DeviceNodeWrapper will be used to provide a device view of current node.

### CODE REVIEW

**Review by Cpaulyz — CHANGES_REQUESTED** [review]
PTAL

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/mnode/impl/CachedMeasurementMNode.java [hunk 1]
```diff
@@ -34,7 +34,7 @@ public class CachedMeasurementMNode extends AbstractMeasurementMNode<ICachedMNod
   public CachedMeasurementMNode(
       IDeviceMNode<ICachedMNode> parent, String name, IMeasurementSchema schema, String alias) {
     super(
-        new CachedBasicMNode(parent == null ? null : parent.getAsMNode(), name),
+        new CachedBasicInternalMNode(parent == null ? null : parent.getAsMNode(), name),
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/mnode/impl/CachedMeasurementMNode.java]: I misunderstood the usage. Already rollback.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/mnode/impl/CachedBasicInternalMNode.java [hunk 2]
```diff
@@ -181,11 +164,31 @@ public void setChildren(IMNodeContainer<ICachedMNode> children) {
    */
   @Override
   public int estimateSize() {
-    return 8 + 80 + 192 + super.estimateSize();
+    return 8
+        + 80
+        + 192
+        + super.estimateSize()
+        + 8
+        + (deviceInfo == null ? 0 : deviceInfo.estimateSize());
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/mnode/impl/CachedBasicInternalMNode.java:176]: Added.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/mnode/impl/CachedBasicInternalMNode.java [hunk 3]
```diff
@@ -181,11 +164,31 @@ public void setChildren(IMNodeContainer<ICachedMNode> children) {
...
+  }
+
+  @Override
+  public MNodeType getMNodeType(Boolean isConfig) {
```
[the remark this answers is not in the document]
Comment (MarcosZyk) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/schemaengine/schemaregion/mtree/impl/pbtree/mnode/impl/CachedBasicInternalMNode.java]: Removed.

**Pull request merged**